In [1]:
!pip install pandas
!pip install mysql-connector

In [2]:
import pandas as pd
import mysql.connector
pd.set_option('display.max_columns',None)

Copyright Jana Schaich Borg/Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)

# MySQL Exercise 8: Joining Tables with Outer Joins

We focused on inner joins in the last set of exercises.  Most of the time inner joins will give you the results you are looking for.  Occasionally, though, you might want to include all the data from a table in your calculations or your output, even if those data do not all match up with the data from the other tables you are joining with.  
    
For example, if you have a table with customer demographic information and a table with information about which customers were sent a free sample, your might want to analyze the characteristics of customers who did and did not receive the sample.  The best way to do that in a program like Tableau would be to have all your customer information in one table with an extra column indicating whether the customer received the free sample or not.  Alternatively, you might want to generate a list of customers who did *not* receive the free sample, so that you can arrange for the customers to receive a free sample in the future.

In these types of situations, you will use outer joins to connect tables.  Outer joins include left joins, right joins, or full outer joins (recall that full outer joins are NOT supported in MySQL).  Refer to the videos "What are Joins?" and "Joins With Many to Many Relationships and Duplicates" for more information about joins. 

Here's a picture to remind you of the general concepts behind outer joins:

<img src="https://duke.box.com/shared/static/4s89rmm8a75tep4puyqt1tdlhic0stzx.jpg" width=400 alt="SELECT FROM WHERE" />

To begin practicing outer joins, load the sql library, connect to the Dognition database, and make the Dognition database your default database: 


In [3]:
%load_ext sql
%sql mysql://studentuser:studentpw@localhost/dognitiondb
%sql USE dognitiondb

 * mysql://studentuser:***@localhost/dognitiondb
0 rows affected.


[]

In [4]:
mydb = mysql.connector.connect(
  host="localhost",
  user="studentuser",
  passwd="studentpw",
  database="dognitiondb"
)

## Left and Right Joins

Left and right joins use a different sytax than we used in the lesson about inner joins.  The method I showed you to execute inner joins tells the database how to relate tables in a WHERE clause like this:

```mySQL
WHERE d.dog_guid=r.dog_guid
```

I find this syntax -- called the "equijoin" syntax -- to be very intuitive, so I thought it would be a good idea to start with it.  However, we can re-write the inner joins in the same syntax used by outer joins.  To use this more traditional syntax, you have to tell the database how to connect the tables using an ON clause that comes right after the FROM clause.  Make sure to specify the word "JOIN" explicitly.  This traditional version of the syntax frees up the WHERE clause for other things you might want to include in your query.  Here's what one of our queries from the inner join lesson would look like using the traditional syntax:

```mySQL
SELECT d.dog_guid AS DogID, d.user_guid AS UserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM dogs d JOIN reviews r
  ON d.dog_guid=r.dog_guid AND d.user_guid=r.user_guid
GROUP BY d.user_guid
HAVING NumRatings > 9
ORDER BY AvgRating DESC
LIMIT 200
```

You could also write "INNER JOIN" instead of "JOIN" but the default in MySQL is that JOIN will mean inner join, so including the word "INNER" is optional.

If you need a WHERE clause in the query above, it would go after the ON clause and before the GROUP BY clause.

Here's an example of a different query we used in the last lesson that employed the equijoin syntax:

```mySQL
SELECT d.user_guid AS UserID, d.dog_guid AS DogID, 
       d.breed, d.breed_type, d.breed_group
FROM dogs d, complete_tests c
WHERE d.dog_guid=c.dog_guid AND test_name='Yawn Warm-up';
```

**Question 1: How would you re-write this query using the traditional join syntax?**

In [5]:
pd.read_sql_query("SELECT d.user_guid AS UserID, d.dog_guid AS DogID, d.breed, d.breed_type, d.breed_group \
                   FROM dogs d JOIN complete_tests c \
                   ON d.dog_guid=c.dog_guid \
                   WHERE test_name='Yawn Warm-up';", mydb)

UserID  \
0      ce135e14-7144-11e5-ba71-058fbc01cf0b   
1      ce135f2c-7144-11e5-ba71-058fbc01cf0b   
2      ce135e14-7144-11e5-ba71-058fbc01cf0b   
3      ce136ac6-7144-11e5-ba71-058fbc01cf0b   
4      ce13615c-7144-11e5-ba71-058fbc01cf0b   
...                                     ...   
20840  ce9b33b6-7144-11e5-ba71-058fbc01cf0b   
20841  ce9b33b6-7144-11e5-ba71-058fbc01cf0b   
20842  ce9acbb0-7144-11e5-ba71-058fbc01cf0b   
20843  ce9aeb5e-7144-11e5-ba71-058fbc01cf0b   
20844  ce9b35e6-7144-11e5-ba71-058fbc01cf0b   

                                      DogID                breed  \
0      fd27b86c-7144-11e5-ba71-058fbc01cf0b             Shih Tzu   
1      fd27bbbe-7144-11e5-ba71-058fbc01cf0b                Mixed   
2      fd27ba1a-7144-11e5-ba71-058fbc01cf0b             Shih Tzu   
3      fd27c5be-7144-11e5-ba71-058fbc01cf0b  Shih Tzu-Poodle Mix   
4      fd27b948-7144-11e5-ba71-058fbc01cf0b       Siberian Husky   
...                                     ...                  ...   
20840  fdc23e96-7144-11e5-ba71-058fbc01cf0b    Yorkshire Terrier   
20841  fdc23e96-7144-11e5-ba71-058fbc01cf0b    Yorkshire Terrier   
20842  fdc1d0a0-7144-11e5-ba71-058fbc01cf0b               Beagle   
20843  fdc1f788-7144-11e5-ba71-058fbc01cf0b    Yorkshire Terrier   
20844  fdc2417a-7144-11e5-ba71-058fbc01cf0b         Bichon Frise   

                             breed_type   breed_group  
0                            Pure Breed           Toy  
1      Mixed Breed/ Other/ I Don't Know          None  
2                            Pure Breed           Toy  
3                           Cross Breed          None  
4                            Pure Breed       Working  
...                                 ...           ...  
20840                        Pure Breed           Toy  
20841                        Pure Breed           Toy  
20842                        Pure Breed         Hound  
20843                        Pure Breed           Toy  
20844                        Pure Breed  Non-Sporting  

[20845 rows x 5 columns]

Now that we've seen the join syntax, we can begin practicing outer joins. Our Dognition data set will make outer joins more challenging than usual, due to the lack of declared primary keys in the original database, the many-to-many relationships, and the presence of duplicate rows and NULL values in columns we will be using to combine tables.  Mastering outer joins in this challenging context, though, will ensure that you understand the fundamental concepts behind joins, which will be a terrific benefit when you start writing queries in other company databases.

The first query I described above was originally written to address the question of whether dog owners who are particularly surprised by their dog's performance on Dognition tests tend to own similar breeds of dogs. When we designed this query in the last lesson, we wanted to focus on the dog owners who reported the highest average amount of surprise at their dog's performance, and provided at least 10 ratings for one or more of their dogs in the ratings.  We also wanted to examine the breed, breed_type, and breed_group of each of these owner's dogs.  

In examining the query, we learned that:

+ All of the user_guids in the reviews table are in the dogs table
+ Only 389 of the over 5000 dog_guids in the reviews table are in the dogs table

The inner join we executed resulted in 389 rows of output, because it only included the data from rows that have equivalent values in both tables being joined.  But what if we wanted the full list of dogs in the reviews table and an indication of whether or not they were in the dogs table, rather than only a list of review information from dogs in the dogs table?  To achieve this list, we could execute an outer join.  
     
Let's start by using a left outer join to get the list we want. When we use the traditional join syntax to write inner joins, the order you enter the tables in your query doesn't matter. In outer joins, however, the order matters a lot. A left outer join will include all of the rows of the table to the left of the LEFT JOIN keywords. A right outer join will include all of the rows of the table to the right of the RIGHT JOIN keywords. So in order to retrieve a full list of dogs who completed at least 10 tests in the reviews table, and include as much breed information as possible, we could query:

```mysql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM reviews r LEFT JOIN dogs d
  ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid
WHERE r.dog_guid IS NOT NULL
GROUP BY r.dog_guid
HAVING NumRatings >= 10
ORDER BY AvgRating DESC;
```

**Question 2: How could you retrieve this same information using a RIGHT JOIN?**


In [6]:
pd.read_sql_query("SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, \
                   d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, \
                   d.breed, d.breed_group, d.breed_type \
                   FROM dogs d RIGHT JOIN reviews r \
                   ON d.dog_guid=r.dog_guid AND d.user_guid=r.user_guid \
                   WHERE r.dog_guid IS NOT NULL \
                   GROUP BY r.dog_guid \
                   HAVING NumRatings >= 10 \
                   ORDER BY AvgRating DESC;", mydb)

rDogID  \
0    fdbf39f8-7144-11e5-ba71-058fbc01cf0b   
1    ce47553e-7144-11e5-ba71-058fbc01cf0b   
2    ce6f07e6-7144-11e5-ba71-058fbc01cf0b   
3    ce45ae5a-7144-11e5-ba71-058fbc01cf0b   
4    ce2a68ac-7144-11e5-ba71-058fbc01cf0b   
..                                    ...   
927  ce81e956-7144-11e5-ba71-058fbc01cf0b   
928  ce8856b0-7144-11e5-ba71-058fbc01cf0b   
929  ce8b700c-7144-11e5-ba71-058fbc01cf0b   
930  ce8c3ee2-7144-11e5-ba71-058fbc01cf0b   
931  ce916868-7144-11e5-ba71-058fbc01cf0b   

                                   dDogID  \
0    fdbf39f8-7144-11e5-ba71-058fbc01cf0b   
1                                    None   
2                                    None   
3                                    None   
4                                    None   
..                                    ...   
927                                  None   
928                                  None   
929                                  None   
930                                  None   
931                                  None   

                                  rUserID  \
0    ce987914-7144-11e5-ba71-058fbc01cf0b   
1    ce6ca9ba-7144-11e5-ba71-058fbc01cf0b   
2    ce7091e2-7144-11e5-ba71-058fbc01cf0b   
3    ce67562c-7144-11e5-ba71-058fbc01cf0b   
4    ce2a45c0-7144-11e5-ba71-058fbc01cf0b   
..                                    ...   
927  ce7f4bf6-7144-11e5-ba71-058fbc01cf0b   
928  ce848d8c-7144-11e5-ba71-058fbc01cf0b   
929  ce87b7f0-7144-11e5-ba71-058fbc01cf0b   
930  ce8862b8-7144-11e5-ba71-058fbc01cf0b   
931  ce88fda4-7144-11e5-ba71-058fbc01cf0b   

                                  dUserID  AvgRating  NumRatings       breed  \
0    ce987914-7144-11e5-ba71-058fbc01cf0b     8.0000          12  Canaan Dog   
1                                    None     7.8750          16        None   
2                                    None     7.5000          10        None   
3                                    None     7.3529          17        None   
4                                    None     7.1333          15        None   
..                                    ...        ...         ...         ...   
927                                  None     0.0000          13        None   
928                                  None     0.0000          21        None   
929                                  None     0.0000          12        None   
930                                  None     0.0000          15        None   
931                                  None     0.0000          11        None   

    breed_group  breed_type  
0       Herding  Pure Breed  
1          None        None  
2          None        None  
3          None        None  
4          None        None  
..          ...         ...  
927        None        None  
928        None        None  
929        None        None  
930        None        None  
931        None        None  

[932 rows x 9 columns]

Notice in the output of both the left and the right version of the outer join, all the rows that had a dog_guid in the reviews table but did NOT have a matching dog_guid in the dogs table have the word "None" entered in output columns related to the dogs table.  "None", in this case, is Jupyter's way of saying the value is NULL.  This becomes clear when you query a list of only the dog_guids that were NOT in the dogs table:

```mysql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM reviews r LEFT JOIN dogs d
  ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid
WHERE d.dog_guid IS NULL
GROUP BY r.dog_guid
HAVING NumRatings >= 10
ORDER BY AvgRating DESC;
```

**Go ahead and try it yourself (you should get 894 rows in your query):**

In [7]:
pd.read_sql_query("SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, \
                   AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, \
                   d.breed, d.breed_group, d.breed_type \
                   FROM reviews r LEFT JOIN dogs d \
                   ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid \
                   WHERE d.dog_guid IS NULL \
                   GROUP BY r.dog_guid \
                   HAVING NumRatings >= 10 \
                   ORDER BY AvgRating DESC;", mydb)

rDogID dDogID  \
0    ce47553e-7144-11e5-ba71-058fbc01cf0b   None   
1    ce6f07e6-7144-11e5-ba71-058fbc01cf0b   None   
2    ce45ae5a-7144-11e5-ba71-058fbc01cf0b   None   
3    ce2a68ac-7144-11e5-ba71-058fbc01cf0b   None   
4    ce72be0e-7144-11e5-ba71-058fbc01cf0b   None   
..                                    ...    ...   
889  ce81e956-7144-11e5-ba71-058fbc01cf0b   None   
890  ce8856b0-7144-11e5-ba71-058fbc01cf0b   None   
891  ce8b700c-7144-11e5-ba71-058fbc01cf0b   None   
892  ce8c3ee2-7144-11e5-ba71-058fbc01cf0b   None   
893  ce916868-7144-11e5-ba71-058fbc01cf0b   None   

                                  rUserID dUserID  AvgRating  NumRatings  \
0    ce6ca9ba-7144-11e5-ba71-058fbc01cf0b    None     7.8750          16   
1    ce7091e2-7144-11e5-ba71-058fbc01cf0b    None     7.5000          10   
2    ce67562c-7144-11e5-ba71-058fbc01cf0b    None     7.3529          17   
3    ce2a45c0-7144-11e5-ba71-058fbc01cf0b    None     7.1333          15   
4    ce73f7a6-7144-11e5-ba71-058fbc01cf0b    None     7.0000          12   
..                                    ...     ...        ...         ...   
889  ce7f4bf6-7144-11e5-ba71-058fbc01cf0b    None     0.0000          13   
890  ce848d8c-7144-11e5-ba71-058fbc01cf0b    None     0.0000          21   
891  ce87b7f0-7144-11e5-ba71-058fbc01cf0b    None     0.0000          12   
892  ce8862b8-7144-11e5-ba71-058fbc01cf0b    None     0.0000          15   
893  ce88fda4-7144-11e5-ba71-058fbc01cf0b    None     0.0000          11   

    breed breed_group breed_type  
0    None        None       None  
1    None        None       None  
2    None        None       None  
3    None        None       None  
4    None        None       None  
..    ...         ...        ...  
889  None        None       None  
890  None        None       None  
891  None        None       None  
892  None        None       None  
893  None        None       None  

[894 rows x 9 columns]

<img src="https://duke.box.com/shared/static/p2eucjdttai08eeo7davbpfgqi3zrew0.jpg" width=600 alt="SELECT FROM WHERE" />

**Question 3: How would you use a left join to retrieve a list of all the unique dogs in the dogs table, and retrieve a count of how many tests each one completed? Include the dog_guids and user_guids from the dogs and complete_tests tables in your output. (If you do not limit your query, your output should contain 35050 rows.  HINT: use the dog_guid from the dogs table to group your results.)**

In [8]:
pd.read_sql_query("SELECT d.user_guid AS dUserID, c.user_guid AS cUserID, d.dog_guid AS dDogID, \
                   c.dog_guid AS dDogID, count(test_name) \
                   FROM dogs d LEFT JOIN complete_tests c\
                   ON d.dog_guid=c.dog_guid \
                   GROUP BY d.dog_guid \
                   ;", mydb)

dUserID cUserID  \
0      ce134e42-7144-11e5-ba71-058fbc01cf0b    None   
1      ce1353d8-7144-11e5-ba71-058fbc01cf0b    None   
2      ce135ab8-7144-11e5-ba71-058fbc01cf0b    None   
3      ce13507c-7144-11e5-ba71-058fbc01cf0b    None   
4      ce135e14-7144-11e5-ba71-058fbc01cf0b    None   
...                                     ...     ...   
35045  ce9b3708-7144-11e5-ba71-058fbc01cf0b    None   
35046  ce9b376c-7144-11e5-ba71-058fbc01cf0b    None   
35047  ce9b37d0-7144-11e5-ba71-058fbc01cf0b    None   
35048  ce9b3834-7144-11e5-ba71-058fbc01cf0b    None   
35049  ce9b376c-7144-11e5-ba71-058fbc01cf0b    None   

                                     dDogID  \
0      fd27b272-7144-11e5-ba71-058fbc01cf0b   
1      fd27b5ba-7144-11e5-ba71-058fbc01cf0b   
2      fd27b6b4-7144-11e5-ba71-058fbc01cf0b   
3      fd27b79a-7144-11e5-ba71-058fbc01cf0b   
4      fd27b86c-7144-11e5-ba71-058fbc01cf0b   
...                                     ...   
35045  fdc2424c-7144-11e5-ba71-058fbc01cf0b   
35046  fdc242ba-7144-11e5-ba71-058fbc01cf0b   
35047  fdc24328-7144-11e5-ba71-058fbc01cf0b   
35048  fdc243a0-7144-11e5-ba71-058fbc01cf0b   
35049  fdc24404-7144-11e5-ba71-058fbc01cf0b   

                                     dDogID  count(test_name)  
0      fd27b272-7144-11e5-ba71-058fbc01cf0b                21  
1      fd27b5ba-7144-11e5-ba71-058fbc01cf0b                20  
2      fd27b6b4-7144-11e5-ba71-058fbc01cf0b                 2  
3      fd27b79a-7144-11e5-ba71-058fbc01cf0b                11  
4      fd27b86c-7144-11e5-ba71-058fbc01cf0b                31  
...                                     ...               ...  
35045                                  None                 0  
35046                                  None                 0  
35047                                  None                 0  
35048                                  None                 0  
35049                                  None                 0  

[35050 rows x 5 columns]

Sometimes you can get so focused on writing your join statement that you don't pay close attention to the fields and tables you put in your other clauses, especially when you are joining a lot of tables.  Often your query will still run successfully, even if you haven't entered the criteria or grouping clause you intended.  The next question will illustrate how easy it is for this to happen.

**Question 4: Repeat the query you ran in Question 3, but intentionally use the dog_guids from the completed_tests table to group your results instead of the dog_guids from the dogs table. (Your output should contain 17987 rows)**

In [9]:
pd.read_sql_query("SELECT d.user_guid AS dUserID, c.user_guid AS cUserID, d.dog_guid AS dDogID, \
                   c.dog_guid AS dDogID, count(test_name) \
                   FROM dogs d LEFT JOIN complete_tests c\
                   ON d.dog_guid=c.dog_guid \
                   GROUP BY c.dog_guid \
                   ;", mydb)

dUserID cUserID  \
0      ce1362ba-7144-11e5-ba71-058fbc01cf0b    None   
1      ce134e42-7144-11e5-ba71-058fbc01cf0b    None   
2      ce1353d8-7144-11e5-ba71-058fbc01cf0b    None   
3      ce135ab8-7144-11e5-ba71-058fbc01cf0b    None   
4      ce13507c-7144-11e5-ba71-058fbc01cf0b    None   
...                                     ...     ...   
17982  ce9b2aba-7144-11e5-ba71-058fbc01cf0b    None   
17983  ce9b2eb6-7144-11e5-ba71-058fbc01cf0b    None   
17984  ce9b306e-7144-11e5-ba71-058fbc01cf0b    None   
17985  ce9b33b6-7144-11e5-ba71-058fbc01cf0b    None   
17986  ce9b35e6-7144-11e5-ba71-058fbc01cf0b    None   

                                     dDogID  \
0      fd27baec-7144-11e5-ba71-058fbc01cf0b   
1      fd27b272-7144-11e5-ba71-058fbc01cf0b   
2      fd27b5ba-7144-11e5-ba71-058fbc01cf0b   
3      fd27b6b4-7144-11e5-ba71-058fbc01cf0b   
4      fd27b79a-7144-11e5-ba71-058fbc01cf0b   
...                                     ...   
17982  fdc235e0-7144-11e5-ba71-058fbc01cf0b   
17983  fdc23a68-7144-11e5-ba71-058fbc01cf0b   
17984  fdc23ba8-7144-11e5-ba71-058fbc01cf0b   
17985  fdc23e96-7144-11e5-ba71-058fbc01cf0b   
17986  fdc2417a-7144-11e5-ba71-058fbc01cf0b   

                                     dDogID  count(test_name)  
0                                      None                 0  
1      fd27b272-7144-11e5-ba71-058fbc01cf0b                21  
2      fd27b5ba-7144-11e5-ba71-058fbc01cf0b                20  
3      fd27b6b4-7144-11e5-ba71-058fbc01cf0b                 2  
4      fd27b79a-7144-11e5-ba71-058fbc01cf0b                11  
...                                     ...               ...  
17982  fdc235e0-7144-11e5-ba71-058fbc01cf0b                 8  
17983  fdc23a68-7144-11e5-ba71-058fbc01cf0b                 1  
17984  fdc23ba8-7144-11e5-ba71-058fbc01cf0b                20  
17985  fdc23e96-7144-11e5-ba71-058fbc01cf0b                 7  
17986  fdc2417a-7144-11e5-ba71-058fbc01cf0b                 6  

[17987 rows x 5 columns]

This time your query ran successfully, but you retrieved many fewer DogIDs because the GROUP BY clause grouped your results according to the dog_guids in the completed_tests table rather than the dog_guid table.  As a result, even though you implemented your join correctly, all of the dog_guids that were in the dogs table but not in the completed_tests table got rolled up into one row of your output where completed_tests.dogs_guid = NULL.  This is a good opportunity to remind ourselves about the differences between SELECT/GROUP BY and COUNT DISTINCT.  

**Question 5: Write a query using COUNT DISTINCT to determine how many distinct dog_guids there are in the completed_tests table.**

In [10]:
pd.read_sql_query("SELECT COUNT(DISTINCT dog_guid)\
                   FROM complete_tests\
                   ;", mydb)

COUNT(DISTINCT dog_guid)
0                     17986

The result of your COUNT DISTINCT clause should be 17,986 which is one row less than the number of rows you retrieved from your query in Question 4.  That's because COUNT DISTINCT does NOT count NULL values, while SELECT/GROUP BY clauses roll up NULL values into one group.  <mark> If you want to infer the number of distinct entries from the results of a query using joins and GROUP BY clauses, remember to include an "IS NOT NULL" clause to ensure you are not counting NULL values.</mark>
   
These exercises are a good illustration of why it is very helpful to save your queries when you are doing an analysis.  Saving your queries allows you and your team members to double-check your work later.  As you can see, the concepts behind SQL aren't themselves too tricky, but it is easy to make mistakes, especially when your queries get long and more complicated.  

One more situation where joins can cause some confusion is when you have duplicate rows in a table you are joining.  If you ignore what we've discussed about set theory and the way databases compute their joins, the behavior databases exhibit when you have duplicate rows in a joined table will seem utterly baffling.  With this knowledge, though, the behavior will make perfect sense.  Let's walk through what happens.   

**Question 6: We want to extract all of the breed information of every dog a user_guid in the users table owns.  If a user_guid in the users table does not own a dog, we want that information as well.  Write a query that would return this information.  Include the dog_guid from the dogs table, and user_guid from both the users and dogs tables in your output. (HINT: you should get 952557 rows in your output!)**  


In [11]:
pd.read_sql_query("SELECT u.user_guid AS uUserID, d.user_guid AS dUserID, d.dog_guid AS dDogID, d.breed\
                   FROM users u LEFT JOIN dogs d\
                   ON u.user_guid=d.user_guid \
                   ;", mydb)

uUserID  \
0       ce134e42-7144-11e5-ba71-058fbc01cf0b   
1       ce134e42-7144-11e5-ba71-058fbc01cf0b   
2       ce1353d8-7144-11e5-ba71-058fbc01cf0b   
3       ce1353d8-7144-11e5-ba71-058fbc01cf0b   
4       ce135ab8-7144-11e5-ba71-058fbc01cf0b   
...                                      ...   
952552  ce9b3708-7144-11e5-ba71-058fbc01cf0b   
952553  ce9b376c-7144-11e5-ba71-058fbc01cf0b   
952554  ce9b376c-7144-11e5-ba71-058fbc01cf0b   
952555  ce9b37d0-7144-11e5-ba71-058fbc01cf0b   
952556  ce9b3834-7144-11e5-ba71-058fbc01cf0b   

                                     dUserID  \
0       ce134e42-7144-11e5-ba71-058fbc01cf0b   
1       ce134e42-7144-11e5-ba71-058fbc01cf0b   
2       ce1353d8-7144-11e5-ba71-058fbc01cf0b   
3       ce1353d8-7144-11e5-ba71-058fbc01cf0b   
4       ce135ab8-7144-11e5-ba71-058fbc01cf0b   
...                                      ...   
952552  ce9b3708-7144-11e5-ba71-058fbc01cf0b   
952553  ce9b376c-7144-11e5-ba71-058fbc01cf0b   
952554  ce9b376c-7144-11e5-ba71-058fbc01cf0b   
952555  ce9b37d0-7144-11e5-ba71-058fbc01cf0b   
952556  ce9b3834-7144-11e5-ba71-058fbc01cf0b   

                                      dDogID                breed  
0       fd27b272-7144-11e5-ba71-058fbc01cf0b   Labrador Retriever  
1       fd417cac-7144-11e5-ba71-058fbc01cf0b                Mixed  
2       fd27b5ba-7144-11e5-ba71-058fbc01cf0b    Shetland Sheepdog  
3       fd3fb0f2-7144-11e5-ba71-058fbc01cf0b    Shetland Sheepdog  
4       fd27b6b4-7144-11e5-ba71-058fbc01cf0b     Golden Retriever  
...                                      ...                  ...  
952552  fdc2424c-7144-11e5-ba71-058fbc01cf0b  Australian Shepherd  
952553  fdc242ba-7144-11e5-ba71-058fbc01cf0b             Havanese  
952554  fdc24404-7144-11e5-ba71-058fbc01cf0b             Havanese  
952555  fdc24328-7144-11e5-ba71-058fbc01cf0b                Mixed  
952556  fdc243a0-7144-11e5-ba71-058fbc01cf0b   Labrador Retriever  

[952557 rows x 4 columns]

There are only 35050 distinct dog_guids in the dogs table.  Why is the database outputting almost a million rows?  That can't be right.  Let's figure out what is going on.  

**Question 7: Adapt the query you wrote above so that it counts the number of rows the join will output per user_id.  Sort the results by this count in descending order.  Remember that if you include dog_guid or breed fields in this query, they will be randomly populated by only one of the values associated with a user_guid (see MySQL Exercise 6; there should be 33,193 rows in your output).**

In [12]:
pd.read_sql_query("SELECT u.user_guid AS uUserID, d.user_guid AS dUserID, d.dog_guid AS dDogID, \
                   d.breed, count(*) AS numrows \
                   FROM users u LEFT JOIN dogs d\
                   ON u.user_guid=d.user_guid \
                   GROUP BY u.user_guid \
                   ORDER BY numrows DESC \
                   ;", mydb)

uUserID  \
0      ce7b75bc-7144-11e5-ba71-058fbc01cf0b   
1      ce225842-7144-11e5-ba71-058fbc01cf0b   
2      ce2258a6-7144-11e5-ba71-058fbc01cf0b   
3      ce135e14-7144-11e5-ba71-058fbc01cf0b   
4      ce29675e-7144-11e5-ba71-058fbc01cf0b   
...                                     ...   
33188  ce9b364a-7144-11e5-ba71-058fbc01cf0b   
33189  ce9b36ae-7144-11e5-ba71-058fbc01cf0b   
33190  ce9b3708-7144-11e5-ba71-058fbc01cf0b   
33191  ce9b37d0-7144-11e5-ba71-058fbc01cf0b   
33192  ce9b3834-7144-11e5-ba71-058fbc01cf0b   

                                    dUserID  \
0      ce7b75bc-7144-11e5-ba71-058fbc01cf0b   
1      ce225842-7144-11e5-ba71-058fbc01cf0b   
2      ce2258a6-7144-11e5-ba71-058fbc01cf0b   
3      ce135e14-7144-11e5-ba71-058fbc01cf0b   
4      ce29675e-7144-11e5-ba71-058fbc01cf0b   
...                                     ...   
33188                                  None   
33189  ce9b36ae-7144-11e5-ba71-058fbc01cf0b   
33190  ce9b3708-7144-11e5-ba71-058fbc01cf0b   
33191  ce9b37d0-7144-11e5-ba71-058fbc01cf0b   
33192  ce9b3834-7144-11e5-ba71-058fbc01cf0b   

                                     dDogID                    breed  numrows  
0      fd7bfb52-7144-11e5-ba71-058fbc01cf0b                 Shih Tzu   913138  
1      fd423714-7144-11e5-ba71-058fbc01cf0b                 Shih Tzu      442  
2      fd40bd62-7144-11e5-ba71-058fbc01cf0b                 Shih Tzu      320  
3      fd27b86c-7144-11e5-ba71-058fbc01cf0b                 Shih Tzu      130  
4      fd46b014-7144-11e5-ba71-058fbc01cf0b  Labrador Retriever- Mix      110  
...                                     ...                      ...      ...  
33188                                  None                     None        1  
33189  fdc241e8-7144-11e5-ba71-058fbc01cf0b      German Shepherd Dog        1  
33190  fdc2424c-7144-11e5-ba71-058fbc01cf0b      Australian Shepherd        1  
33191  fdc24328-7144-11e5-ba71-058fbc01cf0b                    Mixed        1  
33192  fdc243a0-7144-11e5-ba71-058fbc01cf0b       Labrador Retriever        1  

[33193 rows x 5 columns]

This query told us that user 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b' would be associated with 913,138 rows in the output of the outer join we designed!  Once again, why?  We are going to work with the second user_guid in the output you just generated, 'ce225842-7144-11e5-ba71-058fbc01cf0b', because it would be associated with 442 output rows, and 442 rows are much easier to work with than 913,138.   

**Question 8: How many rows in the *users* table are associated with user_guid 'ce225842-7144-11e5-ba71-058fbc01cf0b'?**

In [13]:
pd.read_sql_query("SELECT COUNT(user_guid) \
                   FROM users\
                   WHERE user_guid='ce225842-7144-11e5-ba71-058fbc01cf0b' \
                   ;", mydb)

COUNT(user_guid)
0                17

There are 17 entries associated with that user_guid in the users table.  If you examine all the columns in the entries, you will see that the rows are exact duplicates of each other.  That's unfortunate, but also something that can happen in real life data sets, especially those from new companies or governmental agencies.

Ok, now...

**Question 9: Examine all the rows in the *dogs* table that are associated with user_guid 'ce225842-7144-11e5-ba71-058fbc01cf0b'?**

In [14]:
pd.read_sql_query("SELECT * \
                   FROM dogs\
                   WHERE user_guid='ce225842-7144-11e5-ba71-058fbc01cf0b' \
                   ;", mydb)

gender birthday                       breed  weight  dog_fixed  \
0     male     2013                    Shih Tzu     190        1.0   
1     male     2013                    Shih Tzu     190        1.0   
2     male     2013                    Shih Tzu     190        1.0   
3     male     2013                    Shih Tzu     190        1.0   
4     male     2013                    Shih Tzu     190        1.0   
5     male     2013                    Shih Tzu     190        1.0   
6     male     2013                    Shih Tzu     190        1.0   
7   female     2013                    Shih Tzu     190        1.0   
8     male     2013                    Shih Tzu       0        1.0   
9     male     2013                    Shih Tzu     190        1.0   
10  female     2013                    Shih Tzu     190        1.0   
11    male     2013                    Shih Tzu     190        0.0   
12  female     2012            Spinone Italiano      50        0.0   
13    male     2013                    Shih Tzu     190        0.0   
14  female     2012                    Shih Tzu     190        0.0   
15    male     2009                    Shih Tzu     190        0.0   
16    male     2010                    Shih Tzu     190        0.0   
17    male     2014                    Shih Tzu     190        0.0   
18    male     2013                    Shih Tzu     190        0.0   
19    male     2014                    Shih Tzu     190        NaN   
20    male     2014                    Shih Tzu     190        0.0   
21    male     2009                    Shih Tzu     190        0.0   
22    male     2013                    Shih Tzu     190        0.0   
23  female     2011  Akita-Airedale Terrier Mix      20        0.0   
24    male     2014          Australian Terrier     150        0.0   
25    male     2014                    Shih Tzu     190        0.0   

    dna_tested          created_at          updated_at  dimension  exclude  \
0          1.0 2013-03-21 19:30:06 2015-05-13 18:01:39   maverick      1.0   
1          1.0 2013-03-30 19:41:08 2015-01-06 16:26:41  stargazer      1.0   
2          1.0 2013-04-08 15:44:08 2015-09-21 18:05:51   einstein      1.0   
3          1.0 2013-05-03 19:56:17 2013-07-25 19:42:39       None      1.0   
4          1.0 2013-05-16 16:03:54 2014-10-20 18:10:48       None      1.0   
5          1.0 2013-05-17 16:03:57 2014-11-05 16:18:29  stargazer      1.0   
6          1.0 2013-05-23 12:29:28 2014-10-06 19:29:11       None      1.0   
7          1.0 2013-07-18 19:16:30 2014-05-28 14:11:10       None      1.0   
8          1.0 2013-07-30 19:39:37 2013-07-30 19:39:37       None      NaN   
9          1.0 2013-08-05 19:50:55 2014-07-20 15:06:18       None      1.0   
10         1.0 2013-07-18 17:44:47 2013-10-31 18:32:39       None      1.0   
11         0.0 2013-09-24 18:09:10 2014-05-28 14:09:36       None      1.0   
12         0.0 2013-09-29 14:56:10 2014-06-13 15:13:14       None      NaN   
13         0.0 2013-10-08 21:24:19 2014-05-28 14:09:05       None      1.0   
14         0.0 2013-12-03 18:24:30 2014-05-28 14:08:56       None      1.0   
15         0.0 2013-12-19 19:21:14 2014-05-28 14:09:49       None      1.0   
16         0.0 2014-04-29 14:12:54 2014-06-13 15:13:15       None      1.0   
17         0.0 2014-05-22 21:05:33 2014-07-10 13:57:17       None      1.0   
18         NaN 2014-06-17 20:15:38 2014-06-30 22:28:42  stargazer      1.0   
19         NaN 2014-07-02 14:04:56 2014-07-02 14:04:56       None      NaN   
20         0.0 2014-05-07 18:38:59 2014-05-28 14:06:17       None      1.0   
21         0.0 2014-05-07 20:11:13 2014-05-28 14:04:53       None      1.0   
22         0.0 2014-05-07 20:32:33 2014-05-28 14:03:55       None      1.0   
23         NaN 2014-10-06 02:46:31 2014-10-06 02:46:31       None      NaN   
24         NaN 2015-04-13 14:40:48 2015-04-13 14:40:48       None      NaN   
25         NaN 2015-04-27 15:29:36 2015-04-27 15:29:36       None      NaN   

     b

You should see there are 26 rows associated with that UserID in the dogs table.  When you examine the dogs table, you see that there are a lot of entries that have "Shih Tzu" in the breed column and "190" in the weight column.  This was Dognition's internal convention for indicating test accounts.  So these dog_guids and user_guids do not represent real data.  Nonetheless, they provide a great example of what happens when you join on fields that have duplicate entries.

Recall the general strategy relational databases use to join tables:

<img src="https://duke.box.com/shared/static/km5c7scvo7u6aexzsy1i73wm28aizme1.jpg" width=400 alt="SELECT FROM WHERE" />


When databases join tables, they output the result of every pair of entries that meet certain criteria in the linking column of one table with the linking column of another table.  Our join statement imposed the criteria that the output should only include pairs whose user_guids matched in the two linking columns.  However, since there were multiple rows that had the same user_guid in the users table, *each one of these rows got paired up with each row in the dogs table that had the same user_guid.*  The result was 442 rows, because 17 (instances of the user_guid in the users table) x 26 (instances of the user_guid in the dogs table) = 442.

Having seen this, perhaps you can now appreciate why some database experts emphasize terminology that differentiates between set *theory* and real database *implementation*.  Database operations, like those that join tables, are based on set theory that assumes there are no duplicate rows in your tables.  In real life, duplicate rows get entered all the time, and it can cause you to misinterpret your queries if you do not understand the consequences.  If you've been impacted enough times by the differences between real and theoretical databases, it makes sense that it would be important to you to use language that clearly distinguishes between theory and real life.

The important things I want you to remember from this example of joins with duplicates are that duplicate rows and table relationships that have table-to-table mappings of greater than 1 have multiplicative effects on your query results, due to the way relational databases combine tables.  If you write queries that aggregate over a lot of joined tables, it can be very difficult to catch issues that output results you don't intend, because the aggregated results will hide clues from you.   To prevent this from happening, I recommend you adopt the following practices:

+ Avoid making assumptions about your data or your analyses.  For example, rather than assume that all the values in a column are unique just because some documentation says they should be, check for yourself!
+ Always look at example outputs of your queries before you strongly interpret aggregate calculations. Take extra care to do this when your queries require joins.  
+ When your queries require multiple layers of functions or joins, examine the output of each layer or join first before you combine them all together. 
+ Adopt a healthy skepticsm of all your data and results.  If you see something you don't expect, make sure you explore it before interpreting it strongly or incorporating it into other analyses.

One more type of join to mention that I discussed in the joins videos is a full outer join. Full outer joins include all of the rows in both tables in an ON clause, regardless of whether there is a value that links the row of one table with a row in the other table.  As with left or right joins, whenever a value in a row does not have a matching value in the joined table, NULLs will be entered for all values in the joined table.

Outer joins are used very rarely.  The most practical application is if you want to export all of your raw data to another program for visualization or analysis. The syntax for outer joins is the same as for inner joins, but you replace the word "inner" with " full outer":

```mysql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM reviews r FULL OUTER JOIN dogs d
  ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid
WHERE r.dog_guid IS NOT NULL
GROUP BY r.dog_guid
ORDER BY AvgRating DESC;
```

<mark>**HOWEVER!  MySQL does not support full outer joins.**</mark>
   
If you wanted to imitate a full outer join in mySQL, you could follow one of the methods described at this website:

http://www.xaprb.com/blog/2006/05/26/how-to-write-full-outer-join-in-mysql/





## Practice outer joining your own tables!¶

**Question 10: How would you write a query that used a *left* join to return the number of distinct user_guids that were in the users table, but not the dogs table (your query should return a value of 2226)?**

In [15]:
pd.read_sql_query("SELECT COUNT(DISTINCT u.user_guid) \
                   FROM users u LEFT JOIN dogs d\
                   ON u.user_guid=d.user_guid \
                   WHERE d.user_guid IS NULL \
                   ;", mydb)

COUNT(DISTINCT u.user_guid)
0                         2226

**Question 11: How would you write a query that used a *right* join to return the number of distinct user_guids that were in the users table, but not the dogs table (your query should return a value of 2226)?**

In [16]:
pd.read_sql_query("SELECT COUNT(DISTINCT u.user_guid) \
                   FROM dogs d RIGHT JOIN users u\
                   ON d.user_guid=u.user_guid \
                   WHERE d.user_guid IS NULL \
                   ;", mydb)

COUNT(DISTINCT u.user_guid)
0                         2226

**Question 12: Use a left join to create a list of all the unique dog_guids that are contained in the site_activities table, but not the dogs table, and how many times each one is entered.  Note that there are a lot of NULL values in the dog_guid of the site_activities table, so you will want to exclude them from your list.  (Hint: if you exclude null values, the results you get will have two rows with words in their site_activities dog_guid fields instead of real guids, due to mistaken entries)**

In [17]:
pd.read_sql_query("SELECT s.dog_guid AS SA_dogs_not_present_in_dogs_table, COUNT(*) AS NumEntries \
                   FROM site_activities s LEFT JOIN dogs d\
                   ON s.dog_guid=d.dog_guid \
                   WHERE d.dog_guid IS NULL AND s.dog_guid IS NOT NULL \
                   GROUP BY SA_dogs_not_present_in_dogs_table \
                   ;", mydb)

SA_dogs_not_present_in_dogs_table  NumEntries
0                        Membership        5587
1                     PortalContent          12

**Practice any other outer joins you are interested in here!**